docs: fix errors and inconsistent terminology etc in the rule-engine language doc (#1940)
emqtt/emq-docs-cn: en_US/data-integration/rule-sql-syntax.md

# Rule Engine SQL Language

> An SQL-based based language is used in EMQX's rules for data extraction, filtering, enriching, and transformation. The rule-engine language has a rich set of built-in functions for doing simple transformations and creating timestamps etc. The language also supports embedding [JQ programs](https://stedolan.github.io/jq/) in expressions, which allows you to do complex data transformations when it is needed.

Reviewer: ```suggestion
EMQX uses an SQL-based language in its rules for data extraction, filtering, enrichment, and transformation. The rule-engine language provides a rich set of built-in functions, enabling you to perform simple transformations and create timestamps, among other operations. The language also supports embedding [JQ programs](https://stedolan.github.io/jq/) within expressions, which allows you to do complex data transformations whenever required.
```

> # Rule Engine SQL Reference

Reviewer: ```suggestion
# Rule SQL Reference
```

  - `clientid`: is the client ID in the metadata

>   - `payload.clientid`: is the client ID in the message payload. All fields in the message payload are stored under `payload`

Reviewer: ```suggestion
  - `payload.clientid`: is the client ID in the message payload. All fields in the message payload are stored under `payload`.
```

>     - the `as` syntax renames the `payload.clientid` field as `myclientid`. 

Reviewer: ```suggestion
    - The `as` syntax renames the `payload.clientid` field as `myclientid`. 
```

#### Select by Events

You can also attach rules to events. For example, if you want to get the IP address and port number when client `c1` initiates a connection to EMQX, you can use the statement below:

```sql
SELECT peername as ip_port FROM "$events/client_connected" WHERE clientid = 'c1'
```

::: tip

> You can find all available events in the EMQX Dashboard (under the Events tab when editing a rule)

Reviewer: ```suggestion
You can find all available events in the EMQX Dashboard (under the **Events** tab when editing a rule).
```

SELECT (payload.integer_field + 2) * 2 as num FROM "t/#"
```

You can also use dot notation to access fields in a payload with a complex structure (this assumes that the payload is JSON formatted): 

```sql
SELECT payload.a.b.c.deep as my_field FROM "t/#"
```

> An example of how the equality operator (=) can be used to test that a field has a certain value in the WHERE-clause. Using * in the SELECT-clause will forward all metadata and the payload to the output message:

Reviewer: ```suggestion
The statement below shows an example of how the equality operator (=) can be used to test that a field has a certain value in the WHERE-clause. Using * in the SELECT-clause will forward all metadata and the payload to the output message:
```

```sql
SELECT * FROM "t/#" WHERE payload.x.y = 1
```

> One can use the`and` and `or` operators to form complex boolean expressions in the WHERE-clause:

Reviewer: ```suggestion
You can use the `and` and `or` operators to form complex boolean expressions in the WHERE-clause:
```

[WHERE <condition>]
```

> A `FOREACH` Statement starts with a `FOREACH`-clause that is used to create an array from the input message. The `FROM` and `WHERE` clauses have the same purpose and works the same as the clauses with the corresponding names in the SELECT statement. Besides the `FOREACH`, `FROM` and `WHERE` clauses, the `FOREACH` statement has two more optional clauses:

Reviewer: ```suggestion
A `FOREACH` statement starts with a `FOREACH`-clause that is used to create an array from the input message. The `FROM` and `WHERE` clauses have the same purpose and work the same as the clauses with the corresponding names in the SELECT statement. Besides the `FOREACH`, `FROM` and `WHERE` clauses, the `FOREACH` statement has two more optional clauses:
```
```suggestion
The `FOREACH` statement begins with a `FOREACH` clause, which is used to create an array from the input message. The `FROM` and `WHERE` clauses serve the same purpose and function similarly to their counterparts in the `SELECT` statement. In addition to these clauses, the `FOREACH` statement includes two optional clauses:
```

| Clause   | Optional/Required | Description                                                  |
| -------- | ----------------- | ------------------------------------------------------------ |

> | `DO`     | Optional          | To transform each element in the array selected by `FOREACH`<br><br>Correspond to the `SELECT` clause in the `SELECT` statement and accepts the same expressions |

Reviewer: ```suggestion
| `DO`     | Optional          | Transform each element in the array selected by `FOREACH`.<br><br>Correspond to the `SELECT` clause in the `SELECT` statement and accepts the same expressions. |
```

> | `INCASE` | Optional          | To filter out array elements that do not match the specified conditions.<br><br>Accepts the same expressions as the `WHERE` clause |

Reviewer: ```suggestion
| `INCASE` | Optional          | Filter out array elements that do not match the specified conditions.<br><br>Accepts the same expressions as the `WHERE` clause. |
```

::: tip

> All but the `FOREACH` clause have corresponding clauses in the `SELECT` statement, the `FOREACH` statement can thus be seen as a generalization of the `SELECT` statement as mentioned earlier. The following two statements are equivalent (note that `jq('.', payload)` wraps the payload inside an array): 

Reviewer: ```suggestion
Except for the FOREACH clause, all other clauses in the `FOREACH` statement have corresponding clauses in the `SELECT` statement. This implies that the FOREACH statement can be seen as a generalization of the `SELECT` statement, as mentioned earlier. The following two statements are equivalent (note that `jq('.', payload)` wraps the payload inside an array): 
```

Where,

- The `FOREACH`-clause specifies the `sensors` field in the input message's payload as the array to iterate over. It also names the array itmes `sensor` with the "`as nane`"-syntax.
- The `DO` clause specifies the fields to be included in the output:
  - `timestamp` is the timestamp from the input message's metadata.

>   - `clientid` is the client ID from the input message's metadata, 

Reviewer: ```suggestion
  - `clientid` is the client ID from the input message's metadata.
```

  - `sensor.name` will be capitalized with the build-in `upper` function and renamed as `name` with the "`as name`"-syntax. Please note that "sensor" specifically refers to the current item within the array that is selected by the FOREACH clause.

>   - `sensor.idx`  will be renamed as `idx` with the `as` clause

Reviewer: ```suggestion
  - `sensor.idx`  will be renamed as `idx` with the `as` clause.
```

> - The `INCASE` clause adds another filter condition, so we only get sensors where the value of the idx-field is 1 and above.

Reviewer: ```suggestion
- The `INCASE` clause adds another filter condition, so you only get sensors where the value of the idx-field is 1 and above.
```

> - The `FROM` clause specifies that we want messages from topics that match the pattern "t/#". 

Reviewer: ```suggestion
- The `FROM` clause specifies that you want messages from topics that match the pattern "t/#". 
```

FROM "t/#"
```

> EMQX supports using the `jq` function for advanced transformations, you can refer to the [documentation page for the build-in `jq` function](./rule-sql-jq) for more code examples.

Reviewer: ```suggestion
EMQX supports using the `jq` function for advanced transformations. You can refer to the [ build-in `jq` function](./rule-sql-jq) for more code examples.
```
